src/openzeppelin/tests/test_erc721.cairo

```diff
@@ -0,0 +1,475 @@
+#[should_panic(expected = ('ERC721: invalid token ID', ))]
+fn test__approve_nonexistent() {
+    ERC721::_approve(SPENDER(), TOKEN_ID());
+}
+
+///
+/// set_approval_for_all & _set_approval_for_all
+///
+
+#[test]
+#[available_gas(2000000)]
+fn test_set_approval_for_all_true() {
+    set_caller_address(OWNER());
+    ERC721::set_approval_for_all(OPERATOR(), true);
+    assert(ERC721::is_approved_for_all(OWNER(), OPERATOR()), 'Operator not approved correctly');
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test_set_approval_for_all_false() {
+    set_caller_address(OWNER());
+    ERC721::set_approval_for_all(OPERATOR(), false);
+    assert(!ERC721::is_approved_for_all(OWNER(), OPERATOR()), 'Operator not approved correctly');
+}
+
+#[test]
+#[available_gas(2000000)]
+#[should_panic(expected = ('ERC721: self approval', ))]
+fn test_set_approval_for_all_owner_equal_operator_true() {
+    set_caller_address(OWNER());
+    ERC721::set_approval_for_all(OWNER(), true);
+}
+
+#[test]
+#[available_gas(2000000)]
+#[should_panic(expected = ('ERC721: self approval', ))]
+fn test_set_approval_for_all_owner_equal_operator_false() {
+    set_caller_address(OWNER());
+    ERC721::set_approval_for_all(OWNER(), false);
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test__set_approval_for_all_true() {
+    ERC721::_set_approval_for_all(OWNER(), OPERATOR(), true);
+    assert(ERC721::is_approved_for_all(OWNER(), OPERATOR()), 'Operator not approved correctly');
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test__set_approval_for_all_false() {
+    ERC721::_set_approval_for_all(OWNER(), OPERATOR(), false);
+    assert(!ERC721::is_approved_for_all(OWNER(), OPERATOR()), 'Operator not approved correctly');
+}
+
+#[test]
+#[available_gas(2000000)]
+#[should_panic(expected = ('ERC721: self approval', ))]
+fn test__set_approval_for_all_owner_equal_operator_true() {
+    ERC721::_set_approval_for_all(OWNER(), OWNER(), true);
+}
+
+#[test]
+#[available_gas(2000000)]
+#[should_panic(expected = ('ERC721: self approval', ))]
+fn test__set_approval_for_all_owner_equal_operator_false() {
+    ERC721::_set_approval_for_all(OWNER(), OWNER(), false);
+}
+
+///
+/// transfer_from
+///
+
+#[test]
+#[available_gas(2000000)]
+fn test_transfer_from_owner() {
+    setup();
+
+    assert(ERC721::owner_of(TOKEN_ID()) == OWNER(), 'Ownership before');
+    assert(ERC721::balance_of(OWNER()) == u256_from_felt252(1), 'Balance of owner before');
+    assert(ERC721::balance_of(RECIPIENT()) == u256_from_felt252(0), 'Balance of recipient before');
+
+    set_caller_address(OWNER());
+    ERC721::transfer_from(OWNER(), RECIPIENT(), TOKEN_ID());
+
+    assert(ERC721::owner_of(TOKEN_ID()) == RECIPIENT(), 'Ownership after');
+    assert(ERC721::balance_of(OWNER()) == u256_from_felt252(0), 'Balance of owner after');
+    assert(ERC721::balance_of(RECIPIENT()) == u256_from_felt252(1), 'Balance of recipient after');
+    assert(ERC721::get_approved(TOKEN_ID()).is_zero(), 'Approval not implicitly reset');
```

Comment:
today it's implicitly tested in `test_transfer_from_approved`, but maybe we should have an explicit test with approvals set to check they're properly reset after transfers by owner